gui/options.cpp

```diff
@@ -2156,81 +2208,158 @@ void GlobalOptionsDialog::handleCommand(CommandSender *sender, uint32 cmd, uint3
-		StorageWizardDialog dialog(_selectedStorageIndex);
+	case kSyncSavesStorageCmd: {
+		CloudMan.syncSaves(
+			new Common::Callback<GlobalOptionsDialog, Cloud::Storage::BoolResponse>(this, &GlobalOptionsDialog::storageSavesSyncedCallback)
+		);
+		break;
+	}
+	case kDownloadStorageCmd: {
+		DownloadDialog dialog(_selectedStorageIndex, _launcher);
 		dialog.runModal();
-		//update container's scrollbar
-		reflowLayout();
 		break;
 	}
-	case kRefreshStorageCmd:
-	{
-		CloudMan.info(
-			new Common::Callback<GlobalOptionsDialog, Cloud::Storage::StorageInfoResponse>(this, &GlobalOptionsDialog::storageInfoCallback),
-			new Common::Callback<GlobalOptionsDialog, Networking::ErrorResponse>(this, &GlobalOptionsDialog::storageErrorCallback)
-		);
-		Common::String dir = CloudMan.savesDirectoryPath();
-		if (dir.lastChar() == '/')
-			dir.deleteLastChar();
-		CloudMan.listDirectory(
-			dir,
-			new Common::Callback<GlobalOptionsDialog, Cloud::Storage::ListDirectoryResponse>(this, &GlobalOptionsDialog::storageListDirectoryCallback),
-			new Common::Callback<GlobalOptionsDialog, Networking::ErrorResponse>(this, &GlobalOptionsDialog::storageErrorCallback)
+	case kOpenUrlStorageCmd: {
+		Common::String url = "https://cloud.scummvm.org/";
+		switch (_selectedStorageIndex) {
+		case Cloud::kStorageDropboxId:
+			url += "dropbox";
+			break;
+		case Cloud::kStorageOneDriveId:
+			url += "onedrive";
+			break;
+		case Cloud::kStorageGoogleDriveId:
+			url += "gdrive";
+			break;
+		case Cloud::kStorageBoxId:
+			url += "box";
+			break;
+		}
+
+		if (!g_system->openUrl(url)) {
+			MessageDialog alert(_("Failed to open URL!\nPlease navigate to this page manually."));
+			alert.runModal();
+		}
+		break;
+	}
+	case kPasteCodeStorageCmd: {
+		if (g_system->hasTextInClipboard()) {
+			Common::String message = g_system->getTextFromClipboard();
+			if (!message.empty()) {
+				_storageWizardCodeBox->setEditString(message);
+				_redrawCloudTab = true;
+			}
+		}
+		break;
+	}
+	case kConnectStorageCmd: {
+		Common::String code = "";
+		if (_storageWizardCodeBox)
+			code = _storageWizardCodeBox->getEditString();
+		if (code.size() == 0)
+			return;
+
+		if (CloudMan.isWorking()) {
+			bool cancel = true;
+
+			MessageDialog alert(_("Another Storage is working now. Do you want to interrupt it?"), _("Yes"), _("No"));
+			if (alert.runModal() == GUI::kMessageOK) {
+				if (CloudMan.isDownloading())
+					CloudMan.cancelDownload();
+				if (CloudMan.isSyncing())
+					CloudMan.cancelSync();
+
+				// I believe it still would return `true` here, but just in case
+				if (CloudMan.isWorking()) {
+					MessageDialog alert2(_("Wait until current Storage finishes up and try again."));
+					alert2.runModal();
+				} else {
+					cancel = false;
+				}
+			}
+
+			if (cancel) {
+				return;
+			}
+		}
+
+		if (_storageWizardConnectionStatusHint)
+			_storageWizardConnectionStatusHint->setLabel(_("Connecting..."));
+		CloudMan.connectStorage(
```

Comment:
In my opinion after entering the code and after the account information request completes, the user should be asked to confirm the account is the one he expects before ScummVM starts uploading files to the remote storage.
This is to avoid upload of user files to attacker controlled storage account in case the server is compromised, and to avoid phishing attempts (plz enter this code in ScummVM).

Reply:
Right now we store the auth token for about 10 minutes, allowing users to log in to multiple devices before it expires.

Alternatively, we can change the server implementation to delete the token after the first use in addition to the 10 minutes lifetime.

I've been leaning towards the latter to prevent someone spamming attempts to try and grab a token, but the attack vector is small enough as it is and so low value since we're pretty well scoped.

Reply:
I agree with you regarding the token retrieval endpoint. However I'm not sure how this relates to my point. Perhaps I was not clear enough?

Reply:
Thinking about it some more, once a token is retrieved, it should be deleted from the server.

Reply:
While this is less convinient, I guess that's the price of being more secure. I will do that.

BTW, can anybody share what they are seeing as "username" for OneDrive? I get some gibberish, which is probably how cyrillic is sent by MS. Might also be unicode, which isn't supported by our JSON implementation.

Reply:
Yeah, they send "\uXXXX\uXXXX..." in JSON, which we're parsing as "\u00XX\u00XX..." because AFAIK there is no unicode support in ScummVM. I'm not sure we're OK with that because gibberish instead of username ("display" name, actually) would be confusing.

Reply:
We have `Common::U32String` which maybe could help here?

Reply:
Well... it looks like I've been able to parse the string correctly in json.cpp using `U32String`, and then I use `convertFromU32String()`. I've tried all available encodings, though it probably should be UTF-8, and neither give acceptable results. In UTF-8 and Windows-1251 (default cyrillic) I see something resembeling what is needed (similar to what happens in other applications which handle cyrillic/utf-8 combination poorly). `convertUtf32ToUtf8()` doesn't seem to work either.

What I get is "\x04<correct cyrillic char>\x04<correct cyrillic char>..." in scummvm.ini and just a weird combination of cyrillic chars in GUI. I don't know what will happen if there would be Chinese or Japanise.

Reply:
Yes, the GUI itself doesn't support unicode, which means that you would need to convert to whichever encoding it is using to display it. If translations is enabled, this would be `Common::TranslationManager::getCurrentCharset()` and I guess this could be implemented in `U32String` using the `Common::TranslationManager::getCharsetMapping()` and do a reverse mapping. However this might well fail, for example in your case if you are not using Russian in the ScummVM GUI, it will probably be using an encoding that does not support cyrillic characters. The only set of characters that we are guaranteed to have are the ASCII characters, so we would need to have a fallback that transliterate to ASCII.

@vyzigold is flying through his GSoC tasks and almost finished work on text-to-speech. The next task will be to work on an encoding conversion mechanism, using iconv when possible, in ScummVM. This is something that would be useful for the text to speech (for example to convert from CP850, used by some game engines, to whatever the TTS system needs, usually UTF-8). And it might help here as well. Also @bgK recently mentioned there is something like that in ResidualVM. However while iconv has some support for transliteration, I am not sure it can transliterate cyrillic to ASCII so we might still have issues when the GUI encoding does not support some of the characters used in the string.

Reply:
If I shift unicode cyrillic chars 864 positions down, I get needed chars in GUI. The same constant is stated in Russian article about ISO 8859-5 in Wikipedia: https://ru.wikipedia.org/wiki/ISO_8859-5 (it's not in English version of the article).

Basically, it's enough to use the following code:
```
if (string[i] >= 0x410 && string[i] < 0x450) { // cyrillic
	charsetString += (char)(string[i] - 0x410 + 176);
}
```

When I'm using English, `TransMan.getCurrentCharset()` says it's `ASCII`. When I switch to Russian, it's `iso-8859-5`. But, `TransMan.getCharsetMapping()` returns a very strange table:
```
1f => Я
80000020 => а
```
while in Unicode (and, as it follows from my experiment and that Wikipedia article, in actual ISO 8859-5):
```
42f => Я
430 => а
```
So, there should be continuous interval of uppercase and then lowercase alphabet characters, but TransMan's mapping table maps them in two different intervals and they don't seem to be correct.

I guess we can either leave everything as it is for this PR so someone fixes it later, or I can write a dumb converter for now ("if using ISO 8859-5, shift 864 positions down"), so basically fix it for cyrillic only.

Reply:
The mapping table in the `TranslationManager` seems correct to me and it gives me a mapping from the GUI encoding to unicode. For example for the `ISO 8859-5` it gives me:
```
207 -> 1071 (0x42f)
208 -> 1072 (0x430)
```
And that is indeed a shift by 864.
As I noted to go from unicode to ISO-8859-5 we need to of a reverse mapping (searching for unicode values in the array to get the index (and the index is equal to the character in the current GUI encoding).

However, it is important to note that the `TranslationManager`character mapping tables does not contain the raw unicode values but also has a flag indicating if the character is required (it were designed for loading TTF fonts and check if a font is usable or not for this language). So you need to use a mask of `0x7FFFFFFF` to get the actual unicode value: `TransMan.getCharsetMapping()[207] & 0x7FFFFFFF => 1071`

That being said, in my opinion it would be fine to wait for the work on handling encodings to be done to implement that part cleanly.

Reply:
@bgK what's the concern around server compromise/security in the current implementation? 

If the server is compromised, we have a bigger problem than revocable cloud keys, as an attacker can modify the source for our daily builds.

The phishing vector is so small, and the reward is non-existent (ScummVM saves), as we are properly scoped on all cloud providers to our own files (with the exception of Box, which doesn't have such a scope)

If someone was inclined, they could just get the cleartext refresh token from scummvm.ini

Any hardening should be done server side, but the average user should be presented with a simple workflow to get this going.

Reply:
We can put warnings in both client and server to never trust a key not gotten directly from the website and shorten the TTL of the shortcode to help mitigate.

Reply:
> If the server is compromised, we have a bigger problem than revocable cloud keys, as an attacker can modify the source for our daily builds.

Indeed, modifying downloadable binaries has a much higher value. Compromising the key server could be moderately interesting as it allows to target existing installations.

> The phishing vector is so small, and the reward is non-existent (ScummVM saves).

Well, the cloud sync system also copies files from the remote storage to the user's device. So it would be possible to distribute malware that way (Possibly a specially crafted save file for some engine that achieves local code execution).

> We can put warnings in both client and server to never trust a key not gotten directly from the website.

I'd be curious to know how you would do that.

Reply:
Mataniko means adding text messages on cloud.scummvm.org (like "Don't tell this code to anyone") and in ScummVM GUI (like "Only connect storage if you've got code from cloud.scummvm.org").

I agree that cloud sync technically could be used to deliver payload, but I don't think it should be possible to execute it. Plus I hope it's not really possible to compromise the server in the first place.

So, I'm kinda fine with how it is, but I'm also more or less fine with adding confirmation step. Mataniko says it just complicates the feature for end user while it's already secure enough. Just let me know what you guys decide (=

Reply:
I've added a warning label and a button to manually enable newly connected storage. Until then it does not allow saves sync or files downloading.

This is how storage connection wizard and newly connected storage look like:
![2](https://user-images.githubusercontent.com/1948111/61996647-b21d8800-b0c0-11e9-9823-98145c53c5f4.png)

If we decide not to make users go through this additional step, we can always revert this commit.

Reply:
So @bgK @criezy is there anything else to do?

Reply:
> Storage is yet disabled.

That sounds strange to me. I would suggest "Storage is still disabled." or "Storage is not yet enabled." (the latter sounds better to me than the former).

Or: "You are almost there. Verify that the username is correct and enable the storage".

Reply:
OK, now it's "4. Storage is not yet enabled. Verify that username is correct and enable it:"

Reply:
Sounds good to me. Thanks a lot @Tkachov.